### ISSUE TITLE
Why LOG.old.* is continuously increased for Read Only Mode.

### ISSUE THREAD

**Issue by swyo** [facebook/rocksdb#7966]
This issued is related to #2166 and #7130.

### Right behavior
I know that `keep_log_file_num` can control how many `log.old.*` files to log.

```c++
// read_only_example.cc
#include <string>
#include <assert.h>
#include <rocksdb/db.h>

int main(void){
    rocksdb::Options options;
    options.create_if_missing = true;
    // set keep_log_file_num flag to 1.
    options.keep_log_file_num = 1;
    // write a record to db.
    rocksdb::DB* db;
    rocksdb::Status s = rocksdb::DB::Open(options, "tmp/testdb", &db);
    assert(s.ok());
    s = db->Put(rocksdb::WriteOptions(), "key1", "value");
    assert(s.ok());
    delete db;
    // continuously open db with read only option.
    for (int i = 0; i < 1000; ++i) {
        rocksdb::DB* db;
        rocksdb::Status s = rocksdb::DB::Open(options, "tmp/testdb", &db);
        //rocksdb::Status s = rocksdb::DB::OpenForReadOnly(options, "tmp/testdb", &db);
        assert(s.ok());
        std::string value;
        s = db->Get(rocksdb::ReadOptions(), "key1", &value);
        assert(s.ok());
        assert(value == "value");
        delete db;
    }
    return 0;
}
```

`DB::Open` leads to expected behavior as follows. (If there are many log files, clean it except for `keep_log_file_num` files.)

```shell
testdb (master) $ ls
006840.sst  006876.sst  006882.sst  009029.log  CURRENT  IDENTITY  LOCK  LOG  MANIFEST-009028  OPTIONS-009026  OPTIONS-009031
```

However,   `DB::OpenForReadOnly` doesn't.
To be more specific, continuously opening rockdb with read only option makes `LOG.old.*` files to keep growing forever. 

### Wrong behavior

```c++
// read_only_example.cc
#include <string>
#include <assert.h>
#include <rocksdb/db.h>

int main(void){
    rocksdb::Options options;
    options.create_if_missing = true;
    // set keep_log_file_num flag to 1.
    options.keep_log_file_num = 1;
    ...
    // continuously open db with read only option.
    for (int i = 0; i < 1000; ++i) {
        rocksdb::DB* db;
        //rocksdb::Status s = rocksdb::DB::Open(options, "tmp/testdb", &db);
        rocksdb::Status s = rocksdb::DB::OpenForReadOnly(options, "tmp/testdb", &db);
        assert(s.ok());
        std::string value;
        s = db->Get(rocksdb::ReadOptions(), "key1", &value);
        assert(s.ok());
        assert(value == "value");
        delete db;
    }
    return 0;
}
```

Reproduce the behavior.

<details> <summary> Compile </summary>
<pre><code>
git clone https://github.com/facebook/rocksdb.git rocksdb.git 
cd rocksdb.git && make static_lib
cd ./examples // write read_only_examples.cc, and then add following option in the Makefile.
read_only_example: librocksdb read_only_example.cc
        $(CXX) $(CXXFLAGS) $@.cc -o$@ ../librocksdb.a -I../include -O2 -std=c++11 -g $(PLATFORM_LDFLAGS) $(PLATFORM_CXXFLAGS) $(EXEC_LDFLAGS) 

make read_only_example 
</pre> </code> </details>

<details> <summary> Check the log files </summary>
<pre><code>
// run the executable file.
./read_only_example

// check log files.
tree .
... (생략) ...
|-- LOG.old.1613359644744672
|-- LOG.old.1613359644794876
|-- LOG.old.1613359644843527
|-- LOG.old.1613359644891672
|-- LOG.old.1613359644939172
|-- LOG.old.1613359644987382
|-- LOG.old.1613359645035138
|-- LOG.old.1613359645084536
|-- LOG.old.1613359645133632
|-- LOG.old.1613359645184227
|-- LOG.old.1613359645233189
|-- LOG.old.1613359645283385
|-- LOG.old.1613359645340560
|-- LOG.old.1613359645390889
|-- LOG.old.1613359645442209
|-- LOG.old.1613359645492517
|-- LOG.old.1613359645541744
|-- LOG.old.1613359645592847
|-- LOG.old.1613359645643232
|-- LOG.old.1613359645694900
|-- LOG.old.1613359645745462
|-- LOG.old.1613359645798686
|-- LOG.old.1613359645848691
|-- LOG.old.1613359645899865
|-- MANIFEST-006871
|-- OPTIONS-006866
`-- OPTIONS-006874

0 directories, 1338 files
</pre> </code> </details>

Therefore, I analyzed why this happened. This is my summary.
* General read/write Mode : 
In the [`DB::Open`](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_open.cc#L1451), [DBImpl::Open](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_open.cc#L1456) is called internally, `keep_log_file_num` is used in [`DBImp::PurgeObsoleteFiles`](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_open.cc#L1625), which in in [`DBImp::DeleteObsoleteFiles`](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_open.cc#L1625), so [this code block](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_files.cc#L596-L629) manages the `Log.old.*` files.
* For read only open Mode: 
 In the [`DB::OpenForReadOnly`](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_readonly.cc#L154), CompactedDBImpl](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/compacted_db_impl.h#L14) (which is inherited by `DBImpl`)'s member function [`CompactedDBImpl::Open`](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/db_impl/db_impl_readonly.cc#L165) is called, but the `CompactedDBImpl::Open`'s [code block](https://github.com/facebook/rocksdb/blob/00519187a6e495f0be0bbc666cacd9da467a6c1e/db/compacted_db_impl.cc#L145-L166) does not use `keep_log_file_num` to delete obsolete log files.
 

### Expected Behavior

I want to use `OpenForReadOnly` continuously with`keep_log_file_num` option.  If this problem isn't fixed, the `Log.old. *` will increase monotonically and my disc may become full in the future.

**Comment by riversand963** [comment]
Thanks @swyo for reporting. I am not sure about the design decision behind this. My understanding is that read-only mode will avoid making changes to the db, not only logically but also physically. If we enable file deletions, then users won't be able to open the db in read-only mode if they do not have permission to modify the directory and its files/subdirs. I just did a simple test as follows.
- Create a test db by running `db_bench -benchmarks=fillseq`
- Make the directory readonly so that no new files can be added and no files can be removed. All existing files are readonly too.
- Access the db with ldb. `./ldb scan --db=/tmp/rocksdbtest-148062/dbbench/ --hex scan --max_keys=10 --no_value`. The ldb scan command uses read-only instance, and it works.

If there are such users in this situation, then making the proposed change will certainly be a regression for them. It is not easy for us to track which users have this potential issue. Therefore, I believe we should look for another solution. Two questions.

1. Is there a regular mode database running? If so, we can rely on it to delete old files, as you already noticed.
2. If not, have you considered [secondary instance](https://github.com/facebook/rocksdb/wiki/Secondary-instance) (open via `OpenAsSecondary()`)? It's similar to read-only instance with a few differences. Secondary instance can perform file cleanup in `TryCatchUpWithPrimary()`, which is an API you can call so that the secondary instance will try to catch up with the primary instance. If there is no primary instance, I believe you can still use it. If not, let me know.

**Comment by ajkr** [comment]
> To be more specific, continuously opening rockdb with read only option makes LOG.old.* files to keep growing forever.

Coincidentally @pdillinger just fixed this in #8142, not by cleaning up files in read-only mode, but by not opening a logger by default for read-only DBs. Copied relevant release note here:

> * Fixed some cases in which DB::OpenForReadOnly() could write to the filesystem. If you want a Logger with a read-only DB, you must now set DBOptions::info_log yourself, such as using CreateLoggerFromOptions().

**Comment by swyo** [comment]
Thank you for reading my issue. @riversand963 @ajkr 👍  

> Not by cleaning up files in read-only mode, but by not opening a logger by default for read-only DBs. 

I agree with the solution above for me :)

I'll track down #8142, and I'll check if this fixes the problem.

**Comment by pdillinger** [comment]
I'm assuming fixed with #8142. Re-open this issue if you have trouble.

**Comment by dufferzafar** [comment]
@ajkr I'm using an older RocksDB version, specifically 6.1.2, how can I get this behaviour of not writing a LOG file when opening in ReadOnlyMode.

I see there's a function for `CreateLoggerFromOptions`, but how can I create a logger that won't log anything, or not create any files. Will setting `db_log_dir` to `/dev/null` work?
